recipe/meta.yaml

```diff
@@ -12,7 +21,26 @@ source:
   sha256: {{ sha256 }}
 
 build:
-  number: 4
+  number: {{ build }}
+
+  # add build string so packages can depend on
+  # mpi or nompi variants explicitly:
+  # `pkg * mpi_mpich_*` for mpich
+  # `pkg * mpi_*` for any mpi
+  # `pkg * nompi_*` for no mpi
```

Comment:
Is it possible to have the nompi package have a package name that doesn't reference mpi at all?  It seems more natural that someone will look for the package `moab` than `nompi_moab`

Reply:
I will change pkg to moab in this documentation.

No one had to specify anything about mpi in the package name. The mpi things are in the build string, which is optional to specify. It is important to have a nompi build string because without it there is no way to explicitly ask for the version without mpi though this will be the default because of the higher build number.

Let me know if you have more questions.

Reply:
If you install moab and one of the mpi packages, I think you should get the correct moab package automatically without specifying the build. We will want to test this after the merge.

Reply:
To be more explicit:
```
conda create -n moab_test moab
```
should install moab without MPI
```
conda create -n moab_test moab mpich
```
or
```
conda create -n moab_test moab openmpi
```
should create environments where moab works with the desired MPI flavor.  There might be cases where you want MPI but you don't want `moab` to use it.  In such a case, you could do:
```
conda create -n moab_test moab * nompi_* openmpi
```
to explicitly **not** use MPI in `moab` but to have it available for other packages to use.

A package that depends on `moab` could also use these build string wildcards to explicitly specify which flavor of MPI it wants from `moab` in its recipe.  That is another reason that the `nompi` build string is useful.

Reply:
Hmm, I just tested with `hdf5` and when I do:
```
conda create -n test hdf5 openmpi
```
I nevertheless get the `nompi` version of `hdf5`, so it seems like it's a bit trickier than I thought.

Reply:
```
conda create -n test hdf5=*=mpi_openmpi_* openmpi
```
results in the expected combination.

I'm not sure there's anything we can do about that.  We're following the `conda-forge` protocol here as I think we need to do.  If the resulting behavior isn't what I'd expect, that's up to the `conda-forge` community as a whole to address.

Reply:
I'm more interested in what the set of package names is when I as a user want to `conda install` something?

Reply:
@gonuke, if I understand your comment correctly, `conda create` is the "safer" version of `conda install`.  That is, it is better and safer to create a new conda environment from scratch than it is to install new packages into an existing environment, since upgrades don't always happen cleanly.  That being said, all my examples above apply to `conda install` exactly the same as `conda create`.

Reply:
Maybe I'm being left behind (likely :) ), but I think `conda install` is a [common approach](https://conda-forge.org/#about) for many of the users that I'm interested in supporting, warts and all.  I was imagining a paradigm where users chose one of:
```
conda install -c conda-forge moab
conda install -c conda-forge moab-mpi-mpich
conda install -c conda-forge moab-mpi-openmpi
```
but I'm not sure if that's how it works?

Reply:
Right, the equivalent would be:
```
conda install -c conda-forge moab
conda install -c conda-forge moab=*=mpi_mpich_*
conda install -c conda-forge moab=*=mpi_openmpi_*
```
Instead of installing separate packages (which would be a requirement if `moab-mpi-mpich` or the `openmpi` equivalent were to be made available), we install different builds of the same package.  The first "*" above indicates that we're not specifying a version, so we want the latest compatible version.  The second "*" indicates that we're not constraining the build string beyond the MPI flavor we want, so the largest compatible build number will be selected.

I'm afraid this isn't really up for too much negotiation on conda-forge.  This is the way things are done for MPI here.  But I do hope that makes things clearer at least.

Reply:
Yep - that's all fine - I wasn't looking to change anything, but didn't know the syntax.
Thanks!

Reply:
Thanks @xylar. I like the variations and the flexibility of the new build matrix. Please make sure to document these variants well though so that users do not install the wrong combo inadvertently, especially when propagating MPI dependency to HDF5 (or NetCDF?).